### PR TITLE
[FLINK-35159] Transition ExecutionGraph to RUNNING after slot assignment

### PR DESCRIPTION
**Pull request by zentol**
Transitioning the ExecutionGraph to running implicitly starts periodic checkpoint triggering. This should only occur after the all initialization steps have occurred.
The existing code could leak the CheckpointCoordinator if '#handleExecutionGraphCreation' fails to 'tryToAssignSlots'; this eventually leads to a checkpoint being triggered for old ExecutionGraph for which the operator holders were not initialized yet, which causes an exception that crashes the JM.

~~We now force this to happen in the `Executing` constructor, at which point all the initialization steps must already be complete, and we can be sure that the EG gets transitioned into a terminal state when an error occurs (implicitly disabling checkpointing again).~~

We now do this only after slots have been assigned, right before the transition into Executing. At that point we either transition into executing (where we can be sure we'll transition into a terminal state eventually) or fail hard if an error occurs.

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraphTest.java [hunk 1]
```diff
@@ -166,18 +170,19 @@ static class MockCreatingExecutionGraphContext extends MockStateWithoutExecution
...
+                e -> CreatingExecutionGraph.AssignmentResult.success(e.getExecutionGraph());
 
         private GlobalFailureHandler globalFailureHandler =
                 t -> {
                     // No-op.
                 };
 
         public void setExpectWaitingForResources() {
-            waitingForResourcesStateValidator.expectInput((none) -> {});
+            waitingForResourcesStateValidator.expectInput((none) -> {
```
Comment (dmvk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraphTest.java:184]: uuh, what's happening here 🤔 did we update formatter?

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java [hunk 2]
```diff
@@ -1783,56 +1778,6 @@ void testStopWithSavepointFailsInIllegalState() throws Exception {
                 .withCauseInstanceOf(CheckpointException.class);
     }
 
-    @Test
-    void testSavepointFailsWhenBlockingEdgeExists() throws Exception {
```
Comment (dmvk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java:1787]: do you know why / when this was introduced? it indeed doesn't seem to be related AS; is this tested somewhere else?

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraphTest.java [hunk 3]
```diff
@@ -93,8 +93,12 @@ void testNotPossibleSlotAssignmentTransitionsToWaitingForResources() {
                 ignored -> CreatingExecutionGraph.AssignmentResult.notPossible());
         context.setExpectWaitingForResources();
 
+        final StateTrackingMockExecutionGraph executionGraph = new StateTrackingMockExecutionGraph();
+
         executionGraphWithVertexParallelismFuture.complete(
-                getGraph(new StateTrackingMockExecutionGraph()));
+                getGraph(executionGraph));
+
+        assertThat(executionGraph.getState()).isEqualTo(JobStatus.INITIALIZING);
```
Comment (dmvk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraphTest.java:101]: iiuc, the graph would be running before the change; makes sense

**Review by dmvk — APPROVED** [review]
LGTM 👍 Thanks for digging up the root cause!

I was initially bit worried about the testing part, but it feels that hardening `testNotPossibleSlotAssignmentTransitionsToWaitingForResources` should be enough, because smoke tests should cover that this didn't break anything.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java (hunk 2 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java:1787]: This was added in FLINK-34371 and is covered by tests in the DefaultScheduler:

https://github.com/apache/flink/commit/d4e0084649c019c536ee1e44bab15c8eca01bf13#diff-b4bc1cd606feb86850a18371520b5dd63d02090b8567fdf80730d1d7dd6e693d

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraphTest.java (hunk 3 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraphTest.java:101]: > the graph would be running before the change

yes

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraphTest.java (hunk 1 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraphTest.java:184]: No i screwed up :) My intellij is using a spotless config that isnt fully compatible with Flink...
